### PR TITLE
feat: Update lambda module and bump Terraform/AWS provider versions

### PR DESCRIPTION
**Pull request by bryantbiggs**
## Description
- Update underlying lambda module to latest v2.x which removes `this_` prefix
- Bump min supported Terraform and AWS provider versions to match that of the lambda module (least common denominator)
- Add new workflow file stub for unit tests. This is so that on the next PR the workflow file will be run (they won't run on the first PR introducing them due to security concerns protected by GitHub). This will be used to add in some unit tests for the actual lambda function python code
- Remove the zip file and add a pattern match to the gitignore file - no longer required
- Update pre-commit to latest

## Motivation and Context
- Keep module up to date with other modules
- Prepping project for unit tests for better test integration on pull requests

## Breaking Changes
- No

## How Has This Been Tested?
- [x] I have tested and validated these changes using one or more of the provided `examples/*` projects
	- Tested with the `examples/notify-slack-simple`

### CODE REVIEW

File: main.tf [hunk 1]
```diff
@@ -109,7 +108,7 @@ module "lambda" {
   # the value of presense of KMS. Famous "computed values in count" bug...
   attach_cloudwatch_logs_policy = false
   attach_policy_json            = true
-  policy_json                   = element(concat(data.aws_iam_policy_document.lambda[*].json, [""]), 0)
+  policy_json                   = try(data.aws_iam_policy_document.lambda[0].json, "")
```
Comment (antonbabenko) [inline main.tf:111]: `try()` 🤣

**Review by antonbabenko — APPROVED** [review]

**Comment by antonbabenko** [comment]
This PR is included in version 4.20.0 :tada:

File: main.tf (hunk 1 above)
Comment (bryantbiggs) [reply main.tf:111]: my best friend ... `try()` 🤣

File: outputs.tf [hunk 2]
```diff
@@ -1,4 +1,4 @@
-output "this_slack_topic_arn" {
+output "slack_topic_arn" {
```
Comment (llamahunter) [inline outputs.tf:1]: These are all breaking changes. You should increase the major version number.

**Review by llamahunter — COMMENTED** [review]
breaking changes should have major version number increments.

**Comment by bryantbiggs** [comment]
@llamahunter what breaking change are you seeing?

**Comment by llamahunter** [comment]
> @llamahunter what breaking change are you seeing?

All the names of the outputs changed. Any terraform using this module that referenced those outputs broke with this change.

**Comment by bryantbiggs** [comment]
there was one output changed `this_slack_topic_arn` => `slack_topic_arn` and in terms of Terraform, I wouldn't call it breaking because there are no destructive state changes. is there a small change required on the users end; yes. in hindsight, we probably could have left it in the codebase with a TODO for the next pending breaking change - but in the end it was a small change that is not destructive to provisioned resources

**Comment by llamahunter** [comment]
It's destructive to existing builds. They all break when using `~>4.0` version matching, meaning they were pinning their module usage to backwards compatible API changes only, and only electing to move to a new incompatible version that requires code changes due to API deprecation in a controlled manner.

**Comment by laupow** [comment]
> I wouldn't call it breaking because there are no destructive state changes 

@bryantbiggs Unfortunately, that isn't the whole story. The breaking change was the module output name change.

For example, these errors showed up after running with 4.20.0 +
Screen Shot 2021-12-13 at 3 19 43 PM

Can you at least add a note about renamed variables to the changelog? There's nothing in the changelog that indicates outputs were renamed.

**Comment by llamahunter** [comment]
Technically, per SemVer, they should publish a new 4.x change that rolls back the non-backwards compatible output name change, and label all intervening 4.x builds from 4.20.0 until the fix version as 'dangerous' to use due to the backwards incompatibility.

**Comment by ssp5zone** [comment]
This broke our builds! Had to give a patch fix. 
Please avoid making such changes in the minor release.

image

**Comment by antonbabenko** [comment]
Fixed in #159 , released [v4.24.0](https://github.com/terraform-aws-modules/terraform-aws-notify-slack/releases/tag/v4.24.0).

We will remove `this_slack_topic_arn` in the upcoming major release `v5`.

Once more, apologies that it broke some builds. It was never our intention to do so.

**Pull request merged**